### PR TITLE
add vault store

### PR DESCRIPTION
**Pull request by fe80**
Hi,

This is a proposal, we actually work to use Vault as backend storage.

We eventually need help for review code and add spec.

Regards,

### CODE REVIEW

**Comment by duritong** [comment]
This is awesome and actually I though about doing it for a long time, but never got to the point where vault was required.

Anyway, I would love some more description how trocla keys are ending up in valt.

Also might it be desirable to have a (configurable) global prefix for keys in vault, so they go under their own path?

Also we should probably explain that kv is a mount in vault, which was not obvious to me at the beginning.

I wonder whether we need tests?

**Comment by IAmAStealer** [comment]
I think the main point to upgrade to vault for us might be to include acl to prepare for the future and improve global security. We do not have a main requirement now but it is a good thing to prepare for.
Trocla keys, as I understand are just a pair of key/value. the secret engine kv and kv-v2 work the same so it is pretty easy to work with.
The global prefix can just be the mount point for a start.
But we are thinking how to adapt trocla to allow more options like adding a path for the key, in order to palce them where they belong for acl later on.
Tests are surely required.

File: README.md [hunk 1]
```diff
@@ -298,6 +299,18 @@ store_options:
...
 These examples are by no way complete, moneta has much more to offer. Please have a look at [moneta's documentation](https://github.com/minad/moneta/blob/master/README.md) for further information.
 
+#### Vault backend
+
+For use vault you need to create un kv storage in your vault. The `store_options` are dynamic argument for initializer Vault::Client class (except `:kv`, used to defined the kv name)
+
+```YAML
+store: :vault
+store_options:
+  :kv: secret
```
Comment (dje4om) [inline README.md:309]: :kv could be named :kv_mount for less ambiguity, 'secret' is the default kv v1 mount name in vault, but it's nice to be able to choose the mount we want to use

**Comment by dje4om** [comment]
@fe80 , does the ruby vault lib do abstraction between kv v1 and kv v2 engines ?

**Comment by dje4om** [comment]
> Also might it be desirable to have a (configurable) global prefix for keys in vault, so they go under their own path?

A subpath parameter could be an option to be able to use an existing kv mount but a dedicated kv mount seems fine by default
so the prefix could be the mount `mount_name/secret`, or with subpath `mount_name/subpath/secret`

**Comment by fe80** [comment]
Hi @duritong 

I've complete the readme  with more information about vault. It's better ?

I change the kv options with mount too.

Like say @IAmAStealer , vault can be offer lot of feature require for us (history, acl, native REST API...)

The idea about a default path can be a good, we thinking about this feature too an why not another feature with subpath options to additional the key name. I propose to open an issue about this feature an add on a different merge request ?

For us it's just a poc for this moment, we need to think about the data migration an we probably run to double trocla for puppet (use vault just for put data and test performance).

The kv v1 work but add a path `data`, example with `trocla create 'my/path/key' plain`
* v1
Screenshot_2020-10-23 Vault(1)

* v2
Screenshot_2020-10-23 Vault

**Comment by duritong** [comment]
Ui, I absolutely forgot about this here.

@fe80 do you see anything left?

**Comment by fe80** [comment]
Sorry I don't have see your reply.

We are on testing migration actually and It's works correctly for this time

**Comment by duritong** [comment]
thank you to confirm, I'll merge and guess will release a 0.4.0 soon

**Pull request merged**